aea/cli/utils/package_utils.py

```diff
@@ -0,0 +1,377 @@
+# -*- coding: utf-8 -*-
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Module with package utils of the aea cli."""
+
+import os
+import re
+import shutil
+from pathlib import Path
+from typing import Dict, Optional
+
+import click
+
+from jsonschema import ValidationError
+
+import yaml
+
+from aea import AEA_DIR
+from aea.cli.utils.constants import NOT_PERMITTED_AUTHORS
+from aea.cli.utils.context import Context
+from aea.cli.utils.loggers import logger
+from aea.configurations.base import (
+    AgentConfig,
+    DEFAULT_AEA_CONFIG_FILE,
+    PackageType,
+    PublicId,
+    _get_default_configuration_file_name_from_type,
+)
+from aea.configurations.loader import ConfigLoader
+from aea.crypto.helpers import (
+    IDENTIFIER_TO_KEY_FILES,
+    _try_validate_private_key_path,
+    create_private_key,
+)
+from aea.crypto.registry import registry
+
+
+def _verify_or_create_private_keys(ctx: Context) -> None:
+    """
+    Verify or create private keys.
+
+    :param ctx: Context
+    """
+    path = Path(DEFAULT_AEA_CONFIG_FILE)
+    agent_loader = ConfigLoader("aea-config_schema.json", AgentConfig)
+    fp = path.open(mode="r", encoding="utf-8")
+    aea_conf = agent_loader.load(fp)
+
+    for identifier, _value in aea_conf.private_key_paths.read_all():
+        if identifier not in registry.supported_crypto_ids:
+            ValueError("Unsupported identifier in private key paths.")
+
+    for identifier, private_key_path in IDENTIFIER_TO_KEY_FILES.items():
+        config_private_key_path = aea_conf.private_key_paths.read(identifier)
+        if config_private_key_path is None:
+            create_private_key(identifier)
+            aea_conf.private_key_paths.update(identifier, private_key_path)
+        else:
+            try:
+                _try_validate_private_key_path(identifier, private_key_path)
+            except FileNotFoundError:  # pragma: no cover
+                raise click.ClickException(
+                    "File {} for private key {} not found.".format(
+                        repr(private_key_path), identifier,
+                    )
+                )
```

Comment:
Probably not something to address in this PR, but this structure mixes internal logic exceptions with CLI ones. The `FileNotFoundError` should be raised up to the point where the top level CLI component has to handle it. This would isolate both as much as possible.

When working on this type of code I always think how to support both CLI and GUI, and that usually helps to come up with a modular design.

Reply:
Well, in general I agree, it's better to raise ClickException on a first level. But 2 moments: if we gonna raise another exception anyway we have to catch that so usage will grow up from 1 to at least 4 lines.
And the second moment generally it will never be used outside the CLI command. So the changes in p.1 don't worth it.

Reply:
Capturing more exception types improves the semantics. Working only with `ClickException` means that at a higher level you can only log the error but not do any cleanup (if needed) because you don't know anything about the underlying cause.

Regarding the fact it is only used from the CLI command, fair but if the Click API changes or you ever decide to use something else, a lot of files containing business logic (not UI code!) will need to be modified because all layers are coupled.

Reply:
Directly opposite actually - we raise ClickException only if it should be displayed as user-friendly error message. `click` lib catches that and translates to a CLI error message. Otherwise, if it should fail with other exception - it's not caught, it will failed hardly as it should.
And for the cleanup - we have it exactly for ClickException. Take a look at `aea.cli.utils.decorators.clean_after`. This one catches ClickException, cleans up and then re-raise so the error message could reach the user interface.